Firestore/Protos/CMakeLists.txt

```diff
@@ -97,3 +97,57 @@ target_include_directories(
   firebase_firestore_protos_libprotobuf PUBLIC
   ${FIREBASE_SOURCE_DIR}/Firestore/Protos/cpp
 )
+
+set(NANOPB_GENERATOR_DIR ${FIREBASE_BINARY_DIR}/external/src/nanopb/generator/)
+
+# Build nanopb protoc plugin. (Heavily "inspired" by
+# https://github.com/nanopb/nanopb/blob/master/extra/FindNanopb.cmake)
+set(NANOPB_GENERATOR_SRC
+  ${NANOPB_GENERATOR_DIR}/proto/nanopb.proto
+  ${NANOPB_GENERATOR_DIR}/proto/plugin.proto
+)
+set(NANOPB_GENERATOR_PYTHON_SRC)
+foreach (FIL ${NANOPB_GENERATOR_SRC})
```

Comment:
This loop still knows about `${NANOPB_GENERATOR_DIR}/proto`. Why not make `NANOPB_GENERATOR_SRC` be `nanopb plugin` and then get rid of all the `get_filename_component` abstraction in here?

```cmake
foreach (GENERATOR ${NANOPB_GENERATOR_SRC})
  set(output "${NANOPB_GENERATOR_DIR}/proto/${GENERATOR}_pb2.py")
  # ...
endforeach()
```

Possibly also rename to `NANOPB_GENERATORS` or _BASENAMES or something like that too.

Reply:
Short answer: 'cause I blindly copied from nanopb.

Slightly longer answer: nanopb uses these values twice. In once case, it needs the full path name, in the other, only the components. So nanopb chose to store the fullname and extract where necessary. (They could've equivalently stored the components, and built the full path where necessary.)

We only need the values once though. So I've switched it as you suggested.